<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Examples - Cox-nnet</title>
  

  <link rel="shortcut icon" href="../img/favicon.ico">

  
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700' rel='stylesheet' type='text/css'>

  <link rel="stylesheet" href="../css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../css/theme_extra.css" type="text/css" />
  <link rel="stylesheet" href="../css/highlight.css">

  
  <script>
    // Current page data
    var mkdocs_page_name = "Examples";
  </script>
  
  <script src="../js/jquery-2.1.1.min.js"></script>
  <script src="../js/modernizr-2.8.3.min.js"></script>
  <script type="text/javascript" src="../js/highlight.pack.js"></script>
  <script src="../js/theme.js"></script> 

  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
      <div class="wy-side-nav-search">
        <a href=".." class="icon icon-home"> Cox-nnet</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
        <ul class="current">
          
            <li>
    <li class="toctree-l1 ">
        <a class="" href="..">Home</a>
        
    </li>
<li>
          
            <li>
    <ul class="subnav">
    <li><span>Functions</span></li>

        
            
    <li class="toctree-l1 ">
        <a class="" href="../trainCoxMlp/">trainCoxMlp</a>
        
    </li>

        
            
    <li class="toctree-l1 ">
        <a class="" href="../simpleNetArch/">simpleNetArch</a>
        
    </li>

        
            
    <li class="toctree-l1 ">
        <a class="" href="../CVLoglikelihood/">CVLoglikelihood</a>
        
    </li>

        
            
    <li class="toctree-l1 ">
        <a class="" href="../CIndex/">CIndex</a>
        
    </li>

        
            
    <li class="toctree-l1 ">
        <a class="" href="../L2Profile/">L2Profile</a>
        
    </li>

        
            
    <li class="toctree-l1 ">
        <a class="" href="../L2CVProfile/">L2CVProfile</a>
        
    </li>

        
            
    <li class="toctree-l1 ">
        <a class="" href="../L2CVSearch/">L2CVSearch</a>
        
    </li>

        
            
    <li class="toctree-l1 ">
        <a class="" href="../predictNewData/">predictNewData</a>
        
    </li>

        
            
    <li class="toctree-l1 ">
        <a class="" href="../varImportance/">varImportance</a>
        
    </li>

        
            
    <li class="toctree-l1 ">
        <a class="" href="../saveModel/">saveModel</a>
        
    </li>

        
            
    <li class="toctree-l1 ">
        <a class="" href="../loadModel/">loadModel</a>
        
    </li>

        
    </ul>
<li>
          
            <li>
    <li class="toctree-l1 current">
        <a class="current" href="./">Examples</a>
        
            <ul>
            
                <li class="toctree-l3"><a href="#pbc-example">PBC example</a></li>
                
            
                <li class="toctree-l3"><a href="#kirc-example">KIRC example</a></li>
                
                    <li><a class="toctree-l4" href="#interfacing-and-analysis-with-r">Interfacing and analysis with R</a></li>
                
            
            </ul>
        
    </li>
<li>
          
        </ul>
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="..">Cox-nnet</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="..">Docs</a> &raquo;</li>
    
      
    
    <li>Examples</li>
    <li class="wy-breadcrumbs-aside">
      
    </li>
  </ul>
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <h1>Example usages</h1>

<p>This page contains two examples: </p>
<ul>
<li>a primary biliary cirrhosis (PBC) dataset, containing 7 clinical variables (age, sex, edema, bilirubin concentration, albumin concentration, prothrombin time, and disease stage) with 410 patients</li>
<li>kidney renal cell carcinoma (KIRC) dataset using RNASeq gene expression data.  Counts were normalized with DESeq2 and then log-ed (log2 + 1).  </li>
</ul>
<p>The PBC dataset was obtained from the "survival" package in R and the KIRC dataset was obtained from The Cancer Genome Atlas via the Broad Institute.  Both datasets are standardized (mean = 0 and std. dev = 1) and located in the <em>examples</em> folder.  </p>
<p>Plots were done in R rather than Python due to better support for plotting survival data.  </p>
<h2 id="pbc-example">PBC example</h2>
<p>In this example, the L2 parameter is optimized using the <code>L2CVProfile</code> function, which cross validates the training data.  A model is built using the <code>trainCoxMlp</code> function, and evalulated on the hold out test data.  </p>
<pre><code class="python">from cox_nnet import *
import numpy
import sklearn

# load data
x = numpy.loadtxt(fname=&quot;PBC/x.csv&quot;,delimiter=&quot;,&quot;,skiprows=0)
ytime = numpy.loadtxt(fname=&quot;PBC/ytime.csv&quot;,delimiter=&quot;,&quot;,skiprows=0)
ystatus = numpy.loadtxt(fname=&quot;PBC/ystatus.csv&quot;,delimiter=&quot;,&quot;,skiprows=0)

# split into test/train sets
x_train, x_test, ytime_train, ytime_test, ystatus_train, ystatus_test = \
    sklearn.cross_validation.train_test_split(x, ytime, ystatus, 
    test_size = 0.2, random_state = 100)

#Define parameters
model_params = dict(node_map = None, input_split = None)
search_params = dict(method = &quot;nesterov&quot;, learning_rate=0.01, momentum=0.9,
    max_iter=2000, stop_threshold=0.995, patience=1000, patience_incr=2, rand_seed = 123,
    eval_step=23, lr_decay = 0.9, lr_growth = 1.0)
cv_params = dict(cv_seed=1, n_folds=5, cv_metric = &quot;loglikelihood&quot;,
    L2_range = numpy.arange(-4.5,1,0.5))

#cross validate training set to determine lambda parameters
cv_likelihoods, L2_reg_params, mean_cvpl = L2CVProfile(x_train,ytime_train,ystatus_train,
    model_params,search_params,cv_params, verbose=False)
</code></pre>

<p>If you plot the data, you'll hopefully find a nice maximum to choose as the L2 parameter:<br />
<center><img alt="PBC_cv_likelihoods.png" src="../img/PBC_cv_likelihoods.png" /></center></p>
<pre><code class="python">#Build the model
L2_reg = L2_reg_params[numpy.argmax(mean_cvpl)]
model_params = dict(node_map = None, input_split = None, L2_reg=numpy.exp(L2_reg))
model, cost_iter = trainCoxMlp(x_train, ytime_train, ystatus_train, 
    model_params, search_params, verbose=True)
theta = model.predictNewData(x_test)
</code></pre>

<p>One common way to visualize a survival model is to split the test data by median log hazard ratio and plot the two curves:
<center><img alt="PBC_survival_curves.png" src="../img/PBC_survival_curves.png" /></center></p>
<p><br></p>
<h2 id="kirc-example">KIRC example</h2>
<p>In this example, the L2 parameter is optimized by a validation set (a subset of the training data) since cross-validation is computationally expensive.  High throughput datasets should be run using multiple cores or the GPU.  See: http://deeplearning.net/software/theano/tutorial/using_gpu.html</p>
<pre><code class="Python">from cox_nnet import *
import numpy
import sklearn

# load data
x = numpy.loadtxt(fname=&quot;KIRC/log_counts.csv.gz&quot;,delimiter=&quot;,&quot;,skiprows=0)
ytime = numpy.loadtxt(fname=&quot;KIRC/ytime.csv&quot;,delimiter=&quot;,&quot;,skiprows=0)
ystatus = numpy.loadtxt(fname=&quot;KIRC/ystatus.csv&quot;,delimiter=&quot;,&quot;,skiprows=0)

# split into test/train sets
x_train, x_test, ytime_train, ytime_test, ystatus_train, ystatus_test = \
    sklearn.cross_validation.train_test_split(x, ytime, ystatus, test_size = 0.2, random_state = 1)

# split training into optimization and validation sets
x_opt, x_validation, ytime_opt, ytime_validation, ystatus_opt, ystatus_validation = \
    sklearn.cross_validation.train_test_split(x_train, ytime_train, ystatus_train,
    test_size = 0.2, random_state = 1)

# set parameters
model_params = dict(node_map = None, input_split = None)
search_params = dict(method = &quot;nesterov&quot;, learning_rate=0.01, momentum=0.9, 
    max_iter=2000, stop_threshold=0.995, patience=1000, patience_incr=2, 
    rand_seed = 123, eval_step=23, lr_decay = 0.9, lr_growth = 1.0)
cv_params = dict(cv_metric = &quot;loglikelihood&quot;, L2_range = numpy.arange(-2,1,0.5))

#profile log likelihood to determine lambda parameter
likelihoods, L2_reg_params = L2Profile(x_opt,ytime_opt,ystatus_opt,
    x_validation,ytime_validation,ystatus_validation,
    model_params, search_params, cv_params, verbose=False)

numpy.savetxt(&quot;KIRC_likelihoods.csv&quot;, likelihoods, delimiter=&quot;,&quot;)
</code></pre>

<p><center><img alt="KIRC_cindex.png" src="../img/KIRC_cindex.png" /></center></p>
<pre><code>#build model based on optimal lambda parameter
L2_reg = L2_reg_params[numpy.argmax(likelihoods)]
model_params = dict(node_map = None, input_split = None, L2_reg=numpy.exp(L2_reg))
model, cost_iter = trainCoxMlp(x_train, ytime_train, ystatus_train, 
    model_params, search_params, verbose=True)
theta = model.predictNewData(x_test)
</code></pre>

<h3 id="interfacing-and-analysis-with-r">Interfacing and analysis with R</h3>
<p>The preferred programming environment for many bioinformaticians is R.  There are two ways to use Cox-nnet from R.  1) Using the rPython R package to interface with Cox-nnet (see http://rpython.r-forge.r-project.org/) or 2) calling Python scripts as a system call and reading the results in from the command line.  The 2nd method is detailed below:</p>
<pre><code class="R"># Set Theano parameters
PARAMS = &quot;THEANO_FLAGS=mode=FAST_RUN,device=gpu0,floatX=float32 %s&quot;

# Call python script
system(sprintf(PARAMS,&quot;examples/test_KIRC.py&quot;))

# Read in the results on the test set
as.numeric(readLines(&quot;KIRC_theta.csv&quot;)) -&gt; pred
as.numeric(readLines(&quot;KIRC_ytime_test.csv&quot;)) -&gt; time
as.numeric(readLines(&quot;KIRC_ystatus_test.csv&quot;)) -&gt; status

# Plot survival curves
library(survival)
split_r = as.numeric(pred &gt; median(pred))
high_curve&lt;-survfit(formula = Surv(time[split_r == 1],status[split_r == 1]) ~ 1)  
low_curve&lt;-survfit(formula = Surv(time[split_r == 0],status[split_r == 0]) ~ 1) 
plot(high_curve,main=&quot;&quot;, xlab=&quot;Time to event (days)&quot;, ylab=&quot;Probability&quot;, col= &quot;blue&quot;, conf.int=F, lwd=1, xlim=c(0, max(time)))
lines(low_curve, col= &quot;green&quot;, lwd=1, lty=1, conf.int=F)
legend(&quot;topright&quot;, legend=c(&quot;High PI&quot;, &quot;Low PI&quot;), fill=c(&quot;blue&quot;,&quot;green&quot;))

</code></pre>

<p><center><img alt="KIRC_survival_curves.png" src="../img/KIRC_survival_curves.png" /></center></p>
              
            </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
      
        <a href="../loadModel/" class="btn btn-neutral" title="loadModel"><span class="icon icon-circle-arrow-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="http://www.mkdocs.org">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
	  
        </div>
      </div>

    </section>

  </div>

<div class="rst-versions" role="note" style="cursor: pointer">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
      
        <span><a href="../loadModel/" style="color: #fcfcfc;">&laquo; Previous</a></span>
      
      
    </span>
</div>

</body>
</html>
